Mathlib/Topology/MetricSpace/CoveringNumbers.lean

```diff
@@ -0,0 +1,260 @@
+/-
+Copyright (c) 2025 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, Markus Himmel, Lorenzo Luccioli, Alessio Rondelli, Etienne Marion
+-/
+
+import Mathlib.Data.ENat.Lattice
+import Mathlib.Topology.EMetricSpace.Diam
+import Mathlib.Topology.MetricSpace.MetricSeparated
+
+/-!
+# Covering numbers
+
+We define covering numbers of sets in a pseudo-metric space, which are minimal cardinalities of
+`ε`-covers of sets. We also define the packing number, which is the maximal cardinality of
+an `ε`-separated set.
+
+We prove inequalities between these covering and packing numbers.
+
+## Main definitions
+
+* `IsCover`: a set `C` is an `ε`-cover of another set `A` if every point in `A` belongs to a ball
+  with radius `ε` around a point of `C`.
+* `externalCoveringNumber`: the extenal covering number of a set `A` for radius `ε` is the minimal
+  cardinal of an `ε`cover.
+* `coveringNumber`: the covering number(or internal covering number) of a set `A` for radius `ε` is
+  the minimal cardinal of an `ε`cover contained in `A`.
+* `packingNumber`: the packing number of a set `A` for radius `ε` is the maximal cardinal of
+  an `ε`-separated set in `A`.
+
+## Main statements
+
+* `externalCoveringNumber_le_coveringNumber`: the external covering number is less than or equal to
+  the covering number.
+* `packingNumber_two_le_externalCoveringNumber`: the packing number with radius `2 * ε` is
+  less than or equal to the external covering number for `ε`.
+
+## References
+
+* [R. Vershynin, *High-dimensional probability*][vershynin2018high]
+
+-/
+
+open EMetric Set
+open scoped ENNReal
+
+namespace Metric
+
+variable {X : Type*} [PseudoEMetricSpace X] {A B C : Set X} {ε δ : ℝ≥0∞} {x : X}
+
+section IsCover
+
+/-- A set `C` is an `ε`-cover of another set `A` if every point in `A` belongs to a ball with radius
+`ε` around a point of `C`. -/
+def IsCover (C : Set X) (ε : ℝ≥0∞) (A : Set X) : Prop := ∀ a ∈ A, ∃ c ∈ C, edist a c ≤ ε
```

Comment:
Did you consider connecting this to [Metric.thickening](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/MetricSpace/Thickening.html#Metric.thickening)? I'm not suggesting you should use it instead, but probably you can factor some of the proofs in one way or the other.

Also, I would really like to have this definition (and the thickening one) in the context of uniform spaces, but we can live without it for now.